File: Web Survey/resources/js/recommender/distance_metrics/CanberraDistance.js
import AbstractDistanceMetric from "./AbstractDistanceMetric";

export default class CanberraDistance extends AbstractDistanceMetric {
    //TODO: Figure out typical values


    static getMetricName() {
        return "Canberra Distance"
    }

    calculateDistance(track, otherTrack, attributes) {
        const self = this;

        let distanceSum = 0;

        this.attributeKeys.forEach(attribute => {

            let attributeOwn;
            let attributeOther;

            if(attribute === "loudness") {
                attributeOwn = self.fixLoudness(attributes[track][attribute]);
                attributeOther = self.fixLoudness(attributes[otherTrack][attribute])
            } else {
                attributeOwn = attributes[track][attribute];
                attributeOther = attributes[otherTrack][attribute];
            }

            if(attributeOwn === 0 && attributeOther === 0) {
                distanceSum += 0
            } else {
                distanceSum += Math.abs(attributeOwn - attributeOther) /
                    (Math.abs(attributeOwn) + Math.abs(attributeOther));
            }

        });
        let distance;

        try {
            distance = distanceSum;
        } catch(err){
            console.log("Zero division");
            return NaN;
        }
        return distance;
    }

}

File: Experiment Processing/recommender/distance_metrics/canberra_distance.py
from recommender.distance_metrics.abstract_distance_metric import AbstractDistanceMetric


class CanberraDistance(AbstractDistanceMetric):
    """
    Canberra distance, see https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.distance.canberra.html
    for more information.
    """

    def get_name(self):
        return "Canberra Distance"

    def calculate_distance(self, track_a, track_b):

        distance_sum = 0

        track_a_features = track_a.get_metadata()
        track_b_features = track_b.get_metadata()

        for attribute in self.attribute_keys:
            attribute_own = track_a_features[attribute]
            attribute_other = track_b_features[attribute]

            if attribute_own == 0 and attribute_other == 0:
                distance_sum += 0
            else:
                distance_sum += abs(attribute_own - attribute_other) / (abs(attribute_own) + abs(attribute_other))

        return distance_sum
